py/objstr.c

```diff
@@ -2080,6 +2106,11 @@ mp_obj_t mp_obj_new_str_from_vstr(const mp_obj_type_t *type, vstr_t *vstr) {
     ((byte *)o->data)[o->len] = '\0'; // add null byte
     vstr->buf = NULL;
     vstr->alloc = 0;
+    #if MICROPY_PY_BUILTINS_BYTEARRAY
+    if (ret_type == &mp_type_bytearray) {
+        return mp_obj_new_bytearray_by_ref(o->len, (byte *)o->data);
```

Comment:
It doesn't really make sense to construct the str/bytes and then abandon it just to use the data.

```
const byte *data;
if (vstr->len + 1 == vstr->alloc) {
    data = vstr->buf;
} else {
    data = m_renew(char, vstr->buf, vstr->alloc, vstr->len + 1);
}
data[len] = '\0'; // add null byte
vstr->buf = NULL;
vstr->alloc = 0;

#if MICROPY_PY_BUILTINS_BYTEARRAY
if (ret_type == &mp_type_bytearray) {
    return mp_obj_new_bytearray_by_ref(vstr->len, data);
}
#endif
mp_obj_str_t *o = mp_obj_malloc(mp_obj_str_t, type);
o->len = vstr->len;
o->hash = qstr_compute_hash(data, vstr->len);
o->data = data;
return MP_OBJ_FROM_PTR(o);
```